### PR TITLE
Reorder the possible HTTP proxies to use to download Leiningen from GitH...

### PR DESCRIPTION
**Pull request by fgiasson**
...ub. The new sequence is: PowerShell, Curl and then Wget.

Fixes issue #1800 as experienced by multiple users.

### CODE REVIEW

**Comment by winks** [comment]
+echo It is possible that the download failed due to "curl", 
+echo "wget" or "powershell" inability to retreive GitHub's certificate.

should probably be something like

+echo It is possible that the download failed due to "powershell", 
+echo "curl" or "wget" not being able to retrieve GitHub's certificate.

And removing powershell from path doesn't sound like an easy task, and the order is wrong.

I think the cleanest solution would be to add an optional environment variable to override powershell, if it *really* doesn't work.

Sorry if this sounds overly pedantic :)

**Comment by hyPiRion** [comment]
We have the `HTTP_CLIENT` environment variable that you're able to set. It's possible to specify a way to download insecurely (with emphasis on it being insecure) by e.g. setting it to  `call curl --insecure -f -L -o`.

Perhaps making a powershell example of setting that variable would be an idea?

**Comment by winks** [comment]
oh right, forgot about this and that it's supposed to work on windows too.

**Comment by fgiasson** [comment]
@hyPiRion Yeah that is right. However, any idea how to setup `HTTP_CLIENT` to work with the powershell? Particularly since the parameters appears to be in reverse order?

Also, I can't find the code where it says to try to use the `HTTP_CLIENT` environment variable to call `curl` or `wget`. Any pointers?

**Comment by hyPiRion** [comment]
@fgiasson I don't have enough experience with bat/PowerShell to know how that's done unfortunately.

If you wonder about the insecure example, that's just the old insecure code which was removed in d63b2fa40f7f2042c80b1da676667145ca6a1a3c. There isn't any example of how to use `HTTP_CLIENT` as an env variable, which is why I suggested to make one for it.

For reference, the source for `HTTP_CLIENT` usage is here: https://github.com/technomancy/leiningen/blob/772d38342f7beaa06dadb303f76cad77111659d4/bin/lein.bat#L140-L143

**Comment by fgiasson** [comment]
@hyPiRion Thanks for the reference. Yeah, I don't neither for PowerShell. Anybody else as a clue?

**Comment by winks** [comment]
I think it's fine as it is

I just played around with your change merged on windows and it seems to work as expected:

  1. powershell available
    1. works -> fine
    1. doesn't work -> error msg
      1. user does `set HTTP_CLIENT=wget or curl`
        1. -> works -> fine
        1. -> doesn't work -> worst case. maybe the user will figure it out, maybe they will seek help  1. powershell not available
    1. lein.bat tries wget, then curl
      1. works? -> ok
      1. doesn't work? -> see worst case again


So I think the worst case can't really be solved in a nice automated way that's *better* than now.

TLDR: shouldn't be worse, can only better, let's go. (unless I missed something)

**Comment by fgiasson** [comment]
@winks, @hyPiRion,

Check this new iteration: https://github.com/fgiasson/leiningen/commit/9b60ea0332002dcf43d7dc55f96c4c2112c0cc3c

Should be the behavior outlined by @winks. I agree with that proccess.

However, please test that. I am currently getting that PowerShell error:

```
Exception calling " downloadFile " with "2" argument ( s) : "Demand
was aborted : The connection was unexpectedly terminated ".
The character line: 1: 145
+ ... rkCredentials; $client.DownloadFile($a, $f)}
"https://github.com/technomancy/lei ...
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException
```

And I am not sure why.

**Comment by winks** [comment]
@fgiasson Did you change `2.5.2-SNAPSHOT` to `2.5.1` in line 5? because otherwise it can't work.

I just downloaded your HEAD from your branch and just with that change 2.5.1 downloaded perfectly on Win 7 Home Premium SP1.

**Comment by fgiasson** [comment]
@winks nah, that was the actual issue thanks :)

Working perfectly here.

So, is that patch now acceptable to everyone?

**Comment by hyPiRion** [comment]
Yeah, this is super :) Thanks!

**Comment by fgiasson** [comment]
perfect thanks!
